evap/rewards/views.py

```diff
@@ -13,76 +13,54 @@
 from evap.evaluation.models import Semester, UserProfile
 from evap.evaluation.tools import AttachmentResponse, get_object_from_dict_pk_entry_or_logged_40x
 from evap.rewards.exporters import RewardsExporter
-from evap.rewards.forms import RewardPointRedemptionEventForm
+from evap.rewards.forms import RewardPointRedemptionEventForm, RewardPointRedemptionFormSet
 from evap.rewards.models import (
-    NoPointsSelectedError,
-    NotEnoughPointsError,
-    OutdatedRedemptionDataError,
-    RedemptionEventExpiredError,
     RewardPointGranting,
     RewardPointRedemption,
     RewardPointRedemptionEvent,
     SemesterActivation,
 )
-from evap.rewards.tools import grant_eligible_reward_points_for_semester, reward_points_of_user, save_redemptions
-
-
-def redeem_reward_points(request):
-    redemptions = {}
-    try:
-        for key, value in request.POST.items():
-            if key.startswith("points-"):
-                event_id = int(key.rpartition("-")[2])
-                redemptions[event_id] = int(value)
-        previous_redeemed_points = int(request.POST["previous_redeemed_points"])
-    except (ValueError, KeyError, TypeError) as e:
-        raise BadRequest from e
-
-    try:
-        save_redemptions(request, redemptions, previous_redeemed_points)
-        messages.success(request, _("You successfully redeemed your points."))
-    except (
-        NoPointsSelectedError,
-        NotEnoughPointsError,
-        RedemptionEventExpiredError,
-        OutdatedRedemptionDataError,
-    ) as error:
-        status_code = 400
-        match error:
-            case NoPointsSelectedError():
-                error_string = _("You cannot redeem 0 points.")
-            case NotEnoughPointsError():
-                error_string = _("You don't have enough reward points.")
-            case RedemptionEventExpiredError():
-                error_string = _("Sorry, the deadline for this event expired already.")
-            case OutdatedRedemptionDataError():
-                status_code = 409
-                error_string = _(
-                    "It appears that your browser sent multiple redemption requests. You can see all successful redemptions below."
-                )
-            case _:
-                assert_never(type(error))
-
-        messages.error(request, error_string)
-        return status_code
-    return 200
+from evap.rewards.tools import grant_eligible_reward_points_for_semester, redeemed_points_of_user, reward_points_of_user
 
 
 @reward_user_required
 def index(request):
     status = 200
+
+    total_points_available = reward_points_of_user(request.user)
+    events = RewardPointRedemptionEvent.objects.filter(redeem_end_date__gte=datetime.now().date()).order_by("date")
+
+    formset = RewardPointRedemptionFormSet(
+        request.POST or None,
+        initial=[{"event": e, "points": 0, "total_points_available": total_points_available} for e in events],
+        user=request.user,
+    )
```

Comment:
problematic: `total_points_available` is computed outside of the transaction/locked critical section, formset `clean` is computed without database locking, and `save` might run arbitrarily later -- with the possibility that the same points have been redeemed in between.

For this to be safe, we need to ensure that during the update path, all three steps (calculate limit, clean, save) are stored in the transaction while the rows are locked.